### PR TITLE
Fix verify codecgen

### LINKED ISSUE #25103: verify-codecgen and update-codecgen disagree with each other
See, e.g.

- https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25102/kubernetes-pull-test-unit-integration/23972/
- https://github.com/kubernetes/kubernetes/pull/23684#issuecomment-216654698
- #25102

This seems to be blocking multiple PRs from even being tested.

### PR DESCRIPTION
**Pull request by lavalamp**
Fixes #25103.

### CODE REVIEW

**Comment by lavalamp** [comment]
fyi @wojtek-t

**Comment by mml** [comment]
@lavalamp :'(

```
FAILED   ./hack/../hack/verify-codecgen.sh	369s
Verifying ./hack/../hack/verify-codegen.sh
```

**Comment by lavalamp** [comment]
Huh, I ran it multiple times locally. running again...

**Comment by lavalamp** [comment]
Hm. My local system comes up with a different ordering than the remote system. That's probably the culprit.

**Comment by lavalamp** [comment]
I have to go home. @wojtek-t please feel free to take this over if you see this before I get to it. I think the topological sort is not portable.

**Comment by wojtek-t** [comment]
>
I have to go home. @wojtek-t please feel free to take this over if you see this before I get to it. I think the topological sort is not portable.

Why do you think so? It doesn't use anything super-fancy - it's just using some very standard commands.

Regarding the PR itself, this tricky. The files that are generated depends on whether the other ones (they potentially depend on) are generated or not. So we need a well-defined order of generation (the same for verification).

We should do in topological order, because then the gains are the biggest then.

**Comment by lavalamp** [comment]
> Why do you think so?

Because I looked at the test logs and the logs my system uses, and files were processed in a different order. Perhaps because the files aren't deterministically sorted before the tsort, but I haven't verified that yet.

> We should do in topological order, because then the gains are the biggest then.

Nah, in the verify script it doesn't matter. You have to generate every file in the success case anyway, so you might as well just call the main script, like I've changed it to do.

**Comment by mml** [comment]
> Perhaps because the files aren't deterministically sorted before the tsort

My money is here.  Each invocation may be correct, but it's not stable.

**Comment by wojtek-t** [comment]
But it shouldn't matter - the topological sorting is supposed to fix that...

**Comment by mml** [comment]
The tsort is supposed to make sure we do all dependencies of file F before doing F itself.  Consider only 3 files below:

```
B <----
      +--- A
C <----
```

There are two legitimate orders here: [C, B, A] and [B, C, A].  I haven't carefully read the tsort code, but AIUI which one gets generated depends on the order in which the files are emitted from find in the first place.

**Comment by mml** [comment]
...and if that's correct, it should be enough to just do `generated_files=($(find ... | sort))`

**Comment by wojtek-t** [comment]
@mml - yes that's correct, those two orders can be generated by the existing code.

However, what I was trying to say is that both of them should result in the same generated code according to what I know about ugorji. So it seems it's not the case.

>
...and if that's correct, it should be enough to just do generated_files=($(find ... | sort))

Agree. @lavalamp - can you try that (both in update and verify scripts)?

**Comment by lavalamp** [comment]
Yeah, tsort is not deterministic if the things being sorted don't form a tree. That's what I thought was the problem, and I changed it. Actually the problem is the "io" package is confused with the testing type package under go2idl. I guess "io" is used or not depending on go version. Anyway, I fixed that.

**Comment by lavalamp** [comment]
This is working now. @mml or @wojtek-t PTAL

File: hack/verify-codecgen.sh [hunk 1]
```diff
@@ -35,85 +35,27 @@ generated_files=($(
...
-  rm -rf "${_tmpdir}"
-}
-trap cleanup EXIT
-
-# Sort all files in the dependency order.
-number=${#generated_files[@]}
-for (( i=0; i<number; i++ )); do
-  visited[${i}]=false
+for generated_file in ${generated_files[@]}; do
+  cat "${generated_file}" > "${generated_file}.original"
```
Comment (mml) [inline hack/verify-codecgen.sh]: nit: why `cat` instead of `cp`?

File: hack/verify-codecgen.sh [hunk 2]
```diff
@@ -35,85 +35,27 @@ generated_files=($(
...
+    echo "${generated_file} was out of date. Please run hack/update-codecgen.sh. (If you're running locally, this was run for you already.)"
+    ret=1
   fi
-  rm "${base_generated_file}.tmp"
-  popd > /dev/null
+
+  rm -f "${generated_file}.original"
 done
+
+if [[ $ret -ne 0 ]]; then
```
Comment (mml) [inline hack/verify-codecgen.sh]: nit: Seems safe to drop this surrounding `if`.

File: hack/verify-codecgen.sh (hunk 1 above)
Comment (lavalamp) [inline hack/verify-codecgen.sh]: I didn't want to worry about cp barfing if the file already existed.

File: hack/verify-codecgen.sh (hunk 2 above)
Comment (lavalamp) [inline hack/verify-codecgen.sh]: fixed

**Comment by mml** [comment]
The e2e problem was fixed by #25159

**Comment by luxas** [comment]
@k8s-bot e2e test this please github issue: #25159

**Comment by wojtek-t** [comment]
LGTM - thanks!

**Pull request merged**